### PR TITLE
Rewrite ATen ODS code generator.

### PR DESCRIPTION
**Pull request by stellaraccident**
* Based on new op registry and new signature recognition system.
* Deletes prior code generator from previous attempt (moved some of it into this one).
* Renames old generated tablegen source to "Legacy".
* Generates ODS and import rules for most binary and unary arithmetic ops.
* Removes old generated ops and integration tests that were testing details of the prior setup.

### CODE REVIEW

File: frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+  def std_binary_op(self, reg_name, ods_name, op_name):
+    """"Binary"-ops. These ops typically have:
+      - '.Tensor' variant where the second arg is a Tensor
+      - '.Scalar' variant where the second arg is a Scalar
+      - An '.out' variant which contains a final "outref" argument
+    Actual suffixes vary and the rules are set up to match anything
+    that comes in in one of these forms.
+    In addition, most of these have in-place versions, possibly of the
+    '.Tensor' and '.Scalar' variants above.
+    Note that many of these have more than two arguments (i.e. alpha/beta
```
Comment (silvasean) [inline frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py:113]: unclear how having additional arguments isn't load-bearing? How can we ignore part of the signature?

File: frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+    a list of KernelValueConversion constants that make up arg/return specific
+    conversion flags.
+
+    Returns:
+      - An ods dag list of (ods_name, ods_type) tuples
+      - List of (torch_type, [conversion_flag]) for specifying conversions.
+    """
+    # Generate to ods dag list.
+    ods_dag_list = []
+    for i, sigitem in enumerate(siglist):
```
Comment (silvasean) [inline frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py:228]: Any chance the PyTorch signature list format is described somewhere you can link to? At this point in the code (and callers of this function), without specific knowledge of that format, it gets really hard to follow.

File: frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+                                ods_ins=ods_ins,
+                                ods_outs=ods_outs,
+                                traits=["NoSideEffect"])
+    self.impl_emitter.emit_kernel_methods(ods_name,
+                                          reg_record,
+                                          arg_type_flags=arg_type_flags,
+                                          return_type_flags=return_type_flags,
+                                          promote_trailing_out_tensor=True)
+
+  def std_unary_op(self, reg_name, ods_name, op_name):
```
Comment (silvasean) [inline frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py:155]: "std" here has any specific meaning. could we use less overloaded terminology like "ordinary"? Or maybe just call it `binary_op`?

File: frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+
+  def _emit_dag_list_body(self, items):
+    """Emits a dag of (name, type) pairs."""
+    for index, (ods_name, ods_type) in enumerate(items):
+      is_last = index == len(items) - 1
+      ods_namespec = f":${ods_name}" if ods_name else ""
+      self.print(f"{ods_type}{ods_namespec}", end="\n" if is_last else ",\n")
+
+
+class OdsImplEmitter(EmitterBase):
```
Comment (silvasean) [inline frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py:341]: unclear what "impl" is in this context. Also, it doesn't seem related to ODS?

maybe call it "emit interface method implementations"?

File: frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+    "//===----------------------------------------------------------------------===//",
+    "// This file is automatically generated.  Please do not edit.",
+    "// Generated via:",
+    f"//   python -m {_get_main_module_name()}",
+    "//===----------------------------------------------------------------------===//",
+    "",
+    "",
+])
+
+ODS_IMPL_BANNER = "\n".join([
```
Comment (silvasean) [inline frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py:448]: same comment here about "ods impl" confusing naming.

**Review by silvasean — APPROVED** [review]
Looks great!

File: frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py (hunk 1 above)
Comment (stellaraccident) [reply frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py:113]: Let me word-smith it. We process the whole signature but only have special rules around the first two arguments, which are what varies. The rest are constant-ey things and carried verbatim. If that ever doesn't apply, we just fork the offending op and do it manually or as a one off.

File: frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py (hunk 3 above)
Comment (stellaraccident) [reply frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py]: Switched to 'ordinary'. There are variations and they will become important later.

File: frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py (hunk 5 above)
Comment (stellaraccident) [reply frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py:454]: Switched to CCImplEmitter and associated names.

File: frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py (hunk 2 above)
Comment (stellaraccident) [reply frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py:235]: It's a bit endemic to how it operates and is just (for what we care about) kernel name + type list.

What I did is remove the fragile dependence on the (kernel_name, variant), which is not well-defined (the variant is sometimes arbitrary) and reworked the registration keys to be based on the stringified actual signature, same as for PT internally (i.e. 'aten::add(Tensor,Tensor,Scalar)' vs 'aten::add.Tensor'). This is more correct and likely more clear. I also added a did-you-mean lookup for the error message when one doesn't match, because looking in the list is the most common thing I do.

I think this makes the concept more clear without a lot of documentation. Lmk.

File: frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py (hunk 4 above)
Comment (stellaraccident) [reply frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py]: Renamed to CCImplEmitter. This is the C++ Implementation file.
